src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,616 @@
+	if (rc)
+		DS_ERROR(rc, "dfs_release() failed");
+	D_FREE(rec);
+}
+
+static void
+dcache_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_add(&rec->dr_ref, 1);
+	D_DEBUG(DB_TRACE, "increment ref counter of record '%s' from %u to %u", rec->dr_key, oldref,
+		oldref + 1);
+}
+
+static bool
+dcache_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_sub(&rec->dr_ref, 1);
+	D_DEBUG(DB_TRACE, "decrement ref counter of record '%s' from %u to %u", rec->dr_key, oldref,
+		oldref - 1);
+	D_ASSERT(oldref >= 1);
+	return oldref == 1;
+}
+
+static uint32_t
+dcache_rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	return d_hash_string_u32(rec->dr_key, rec->dr_key_len);
+}
+
+static d_hash_table_ops_t dcache_hash_ops = {
+    .hop_key_cmp    = dcache_key_cmp,
+    .hop_rec_addref = dcache_rec_addref,
+    .hop_rec_decref = dcache_rec_decref,
+    .hop_rec_free   = dcache_rec_free,
+    .hop_rec_hash   = dcache_rec_hash,
+};
+
+static int
+dcache_destroy_act(dfs_dcache_t *dcache);
+static int
+dcache_find_insert_act(dcache_rec_t **rec, dfs_dcache_t *dcache, char *path, size_t path_len);
+static void
+drec_incref_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static void
+drec_decref_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static void
+drec_del_at_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static int
+drec_del_act(dfs_dcache_t *dcache, char *path, dcache_rec_t *parent);
+
+static int
+dcache_create_act(dfs_dcache_t **dcache, dfs_t *dfs, uint32_t bits, uint32_t rec_timeout)
+{
+	dfs_dcache_t *dcache_tmp = NULL;
+	dfs_obj_t    *obj        = NULL;
+	daos_obj_id_t obj_id;
+	mode_t        mode;
+	int           rc;
+
+	D_ALLOC_PTR(dcache_tmp);
+	if (dcache_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	dcache_tmp->dd_dfs = dfs;
+
+	rc = dfs_lookup(dfs, "/", O_RDONLY, &obj, &mode, NULL);
+	if (rc != 0)
+		D_GOTO(error, rc = daos_der2errno(rc));
+	rc = dfs_obj2id(obj, &obj_id);
+	D_ASSERT(rc == 0);
+	rc = snprintf(&dcache_tmp->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE,
+		      "%016" PRIx64 "-%016" PRIx64 ":", obj_id.hi, obj_id.lo);
+	D_ASSERT(rc == DCACHE_KEY_PREF_SIZE - 1);
+
+	rc = d_hash_table_create_inplace(D_HASH_FT_MUTEX | D_HASH_FT_LRU, bits, NULL,
+					 &dcache_hash_ops, &dcache_tmp->dd_dir_hash);
+	if (rc != 0)
+		D_GOTO(error, rc);
+
+	dcache_tmp->destroy_fn     = dcache_destroy_act;
+	dcache_tmp->find_insert_fn = dcache_find_insert_act;
+	dcache_tmp->drec_incref_fn = drec_incref_act;
+	dcache_tmp->drec_decref_fn = drec_decref_act;
+	dcache_tmp->drec_del_at_fn = drec_del_at_act;
+	dcache_tmp->drec_del_fn    = drec_del_act;
+
+	*dcache = dcache_tmp;
+	D_GOTO(out, rc = -DER_SUCCESS);
+
+error:
+	D_FREE(dcache_tmp);
+out:
+	dfs_release(obj);
```

Comment:
It was done on purpose as `dfs_release(NULL)` is well supported. The goal is to have one return point from the function: I do not like not mix goto and several return statements.  However this is a personal coding practice, and thus I have no problem to change the GOTO from line 178 to a return statement.

Reply:
Normally the unwinding at the end of the function is done in the opposite order to allocations etc in the main function so that goto can be used to jump to the appropriate place and only unwind the things it needs to.